research/global-config/index.md

```diff
@@ -0,0 +1,361 @@
+
+- let the user know what the final configuration looks like (both for CLI and service).
+
+#### Jinja2 vs Ansible library
+
+For template management I would probably just use the **jinja2 template library**, even though we can also think about the ansible library.
+The ansible library could let us use `built-in filters and functions` but I don't see use cases for them and as a cons it has a heavier dependency footprint.
+
+### Performances
+
+Splitting the configuration in multiple configuration files will lead obviously to worst performance. Personally I don't see a way to prevent it.
+
+We can limit the number of recursion steps; 3/4 steps are, from my point of view, more than enough. Having a recursion limit will avoid an infinite recursion for malformed configurations.
+
+### packit-service defaults
+
+It could happen that the packit-service config defaults for the "Fedora CI instance" and those for the "Usual instance" (as an example) diverge.
+
+If it happens we could create two _hidden_, _inner_ packit config bases, one per instance, which will always be used for merging any packit config we process; in this way the differences will be grouped explicitly in a single place and we could, probably, enable and disable jobs for one instance or the other (as an example `pull-from-upstream` should not appear in fedora ci packit configuration) just using templating.
+
+<!--
+We already have a configuration file that enables packit we can still count on it.
+
+- it could work for all packit instances:
+  - upstream ci + downstream sync experience
+  - downstream ci
+    We already have a mixed experience for upstream ci and downstream sync, from an user point of view it makes no much sense to split the downstream ci configuration because it runs on a different packit service instance.
+    If we want to have a different configuration for the downstream ci probably we should think about splitting configurations for different users experiences?
+  - upstream ci
+  - downstream ci
+  - sync release
+
+Or we can have just one packit config
+
+- by default one project in distgit with the packit config would be enabled the downstream CI and for this reason we should probably have a key to be able to opt-out from it (both as a user action or as a packit configuration action) we can achive it using the above template mechanism and always apply out default to a user configuration.
+  Or **the CI experience will be anabled if the user refers the related global config**.
+  However if the user configuration is against the will of our configuration we should decide which configuration wins (I would say ours)
+- it should be visible which is the result of applying the user packit config to a packit instance config. So we should probably have a command line or a packit service command to show the user the resulting configuration.
+-->
```

Comment:
With any solution, we should provide an easy way to get an expanded configuration. Locally, but maybe also in the logs with clear references where these external configs came from.